### DISCUSSION TITLE
Errors in async iterators cause the websocket to close

### DISCUSSION THREAD

**Discussion by RMHonor** [enisdenjo/graphql-ws#333] [Q&A] [answered]
When running a subscription operation, if the AsyncIterator errors, then it logs `Internal error occurred during message handling...`, and immediately closes the websocket.

I believe this is due to the onMessage handler for subscription operations not having any error handling: https://github.com/enisdenjo/graphql-ws/blob/master/src/server.ts#L799-L811

For example:

```typescript
const resolvers: Resolvers = {
  Subscription: {
    notifications: {
      resolve: (source: any) => {
        if (source instanceof Error) {
          throw source;
        }

        return source;
      },
      subscribe: () => {
        return {
          async *[Symbol.asyncIterator]() {
            await new Promise((resolve) => setTimeout(resolve, 500));
            yield 'Hello world';
            // when this error is thrown, the whole websocket closes
            throw new Error('boom');
          },
        };
      },
    },
  },
};
```

I would've expected throwing the error in `resolve` would attach the error the operationResult.

**Further Information**
graphql-ws version: 5.6.2
node version: v14

**Comment by enisdenjo** [comment]
Are you sure it's the thrown error making problems and not your malformed yield? Please share the full error message.

```diff
const resolvers: Resolvers = {
  Subscription: {
    notifications: {
      resolve: (source: any) => {
        if (source instanceof Error) {
          throw source;
        }

        return source;
      },
      subscribe: () => {
        return {
          async *[Symbol.asyncIterator]() {
            await new Promise((resolve) => setTimeout(resolve, 500));
-           yield 'Hello world';
+           yield { notifications: 'Hello world' };
            // when this error is thrown, the whole websocket closes
            throw new Error('boom');
          },
        };
      },
    },
  },
};
```

**Reply by RMHonor** [reply]
The yield should be correct, the schema is:

```graphql
extend type Subscription {
  notifications: String!
}
```

The error that occurs is:

```bash
Internal error occurred during message handling. Please check your implementation. Error: boom
    at Object.[Symbol.asyncIterator] (/Users/richardhonor/uw/crm-graphql/src/modules/notifications/notifications.resolver.ts:49:19)
    at AsyncGenerator.next (<anonymous>)
    at Object.next (/Users/richardhonor/uw/crm-graphql/node_modules/graphql/subscription/mapAsyncIterator.js:53:23)
    at onMessage (/Users/richardhonor/uw/crm-graphql/node_modules/graphql-ws/lib/server.js:207:165)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at WebSocket.<anonymous> (/Users/richardhonor/uw/crm-graphql/node_modules/graphql-ws/lib/use/ws.js:83:21)
```

I do get an initial message back to the client with "Hello world", so it is sending valid data back:

```json
{
  "id": "8cbd7094-a307-4ae4-be0a-ec14a9b368e7",
  "type": "next",
  "payload": {
    "data": {
      "notifications": "Hello world"
    }
  }
}
```

But then when the error is thrown, the socket connection closes.

**Reply by enisdenjo** [reply]
Well then, if you expect errors thrown from `subscribe` to not bubble but instead change the operation result - the issue lies with [`graphql-js`](https://github.com/graphql/graphql-js) instead.

**Comment by enisdenjo** [comment]
Seems like this is intentional, please see https://github.com/graphql/graphql-js/issues/1539.

**Reply by RMHonor** [reply]
This issue seems to be discussing throwing of errors before the async iterator begins, rather than inside the async iterator itself.

I'm assuming this is the intended behaviour for graphql-js, but I can't see any obvious way of handling errors. I'll have to keep digging.

**Reply by enisdenjo** [reply]
Doesn't really matter when the error is thrown as long as it's thrown from within the `subscribe`.

**Reply by enisdenjo** [reply]
Alternatively you can implement your own server through [`makeServer`]() and handle your errors however you see fit.

Consult the ["Server usage with ws" recipe](https://github.com/enisdenjo/graphql-ws#ws) for a basic implementation and [`use/ws`](https://github.com/enisdenjo/graphql-ws/blob/master/src/use/ws.ts) for a comprehensive one.

**Reply by RMHonor** [reply]
Thanks @enisdenjo, I've got ahead with that approach, and caught the error with the following block in the message handler:

```typescript
socket.on('message', async (event) => {
            try {
              await cb(String(event));
            } catch (err) {
              const msg = parseMessage(String(event));

              // any errors on subscribe messages should be gracefully sent back to the client, not killing the socket
              if (msg.type === MessageType.Subscribe) {
                logger.error({ operation: msg.payload.operationName }, 'operation error');
                socket.send(
                  stringifyMessage<MessageType.Next>({
                    id: msg.id,
                    payload: {
                      errors: [
                        err instanceof GraphQLError
                          ? err
                          : new GraphQLError((err as Error).message),
                      ],
                    },
                    type: MessageType.Next,
                  }),
                );
              } else {
                socket.close(
                  CloseCode.InternalServerError,
                  isProd
                    ? 'Internal server error'
                    : limitCloseReason((err as Error).message, 'Internal server error'),
                );
              }
            }
          }),
```

A minor downside here is that it skips the `onNext()` handler, but it perfectly satisfies our use case!

**Reply by RMHonor** [reply]
I wonder if this might impact more people as they migrate away from `subscriptions-transport-ws`, as it seems there is error handling in their library: https://github.com/apollographql/subscriptions-transport-ws/blob/323e3af83f13512d320125ae2a00be25433fb9d0/src/server.ts#L360-L404

More than happy to raise a PR to handle these errors, if it's something you'd want adding, possibly behind a flag?

**Reply by enisdenjo** [reply]
No, `graphql-ws` will not handle errors thrown from `subscribe` out-of-the box.

As mentioned earlier, the issue lies with [`graphql-js`](https://github.com/graphql/graphql-js) and https://github.com/graphql/graphql-js/issues/1539. You're more than free to open an issue (or push a PR) on their side instead.

Boils down to this: if [`graphql-js`](https://github.com/graphql/graphql-js) deems errors from `subscribe` fatal and bubbles them, so should `graphql-ws`.

**Reply by enisdenjo** [reply]
@RMHonor small remark - you should use the [`Error` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#error) to communicate that the subscribe attempt failed (looking at you're example, that's when you expect errors).

Using the `Error` message instead will error out the subscription and dispose it on the client; whereas using the [`Next` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#next) has the client think that the subscribe _was_ successful and is expecting more "nexts".

```diff
socket.on('message', async (event) => {
            try {
              await cb(String(event));
            } catch (err) {
              const msg = parseMessage(String(event));

              // any errors on subscribe messages should be gracefully sent back to the client, not killing the socket
              if (msg.type === MessageType.Subscribe) {
                logger.error({ operation: msg.payload.operationName }, 'operation error');
                socket.send(
-                 stringifyMessage<MessageType.Next>({
+                 stringifyMessage<MessageType.Error>({
                    id: msg.id,
-                   payload: {
-                     errors: [
-                       err instanceof GraphQLError
-                         ? err
-                         : new GraphQLError((err as Error).message),
-                     ],
-                   },
+                   payload: [
+                      err instanceof GraphQLError
+                        ? err
+                        : new GraphQLError((err as Error).message),
+                   ],
-                   type: MessageType.Next,
+                   type: MessageType.Error,
                  }),
                );
              } else {
                socket.close(
                  CloseCode.InternalServerError,
                  isProd
                    ? 'Internal server error'
                    : limitCloseReason((err as Error).message, 'Internal server error'),
                );
              }
            }
          }),
```

**Reply by RMHonor** [reply]
I thought it should be an `error` type initially, but it seems by default the client responds to an `error` by removing all the listeners, which causes a new websocket for all subsequent operations. I believe that comes from here? https://github.com/enisdenjo/graphql-ws/blob/master/src/client.ts#L552

Are there any issues with returning an execution result with an error on it rather than data from a subscription?

**Reply by enisdenjo** [reply]
The `errorOrClose` you're referencing is listening for WebSocket connection [`onerror`](https://udn.realityripple.com/docs/Web/API/WebSocket/error_event) or [`onclose`](https://udn.realityripple.com/docs/Web/API/WebSocket/close_event) events, **not** the [GraphQL over WebSocket `Error` message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#error).

Handling the `Error` message happens [here](https://github.com/enisdenjo/graphql-ws/blob/762f29e02685b94c9063d817ba2a97e6140b6202/src/client.ts#L857-L862) instead.

---

> Are there any issues with returning an execution result with an error on it rather than data from a subscription?
 
There are no other issues except for those I mentioned in https://github.com/enisdenjo/graphql-ws/discussions/333#discussioncomment-2332628:

> Using the Error message instead will error out the subscription and dispose it on the client; whereas using the [Next message](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#next) has the client think that the subscribe was successful and is expecting more "nexts".

Meaning, if you use the `Next` message and don't dispose of the subscription manually - it will live forever leading to potential memory leaks.

On the other hand, the `Error` message is fatal **only** for the subscription that received the message, not the whole WS connection.

**Reply by RMHonor** [reply]
Thanks a lot for the help. Got ourselves a working solution for now.

Going to dig into graphql-js and see if there's anything we can do with the asynciterator there.

**Reply by elderapo** [reply]
> I wonder if this might impact more people as they migrate away from `subscriptions-transport-ws`, as it seems there is error handling in their library: https://github.com/apollographql/subscriptions-transport-ws/blob/323e3af83f13512d320125ae2a00be25433fb9d0/src/server.ts#L360-L404
> 
> More than happy to raise a PR to handle these errors, if it's something you'd want adding, possibly behind a flag?

This just happened to me. I've been successfully using `subscriptions-transport-ws` but decided to migrate to `graphql-ws` because of the deprecation notices. Now I am considering going back to `subscriptions-transport-ws`...

**Reply by enisdenjo** [reply]
https://github.com/enisdenjo/graphql-ws/discussions/333#discussioncomment-2332493 will keep being the case.

You're free to implement your own server and gracefully handle subscription errors as stated in https://github.com/enisdenjo/graphql-ws/discussions/333#discussioncomment-2332628.

**Reply by elderapo** [reply]
Well, that's going to be kind of tricky because I am using `@nestjs/graphql` which internally uses `graphql-ws` to create the server,

**Reply by enisdenjo** [reply]
Ok, I understand.

Alternatively, you can fix your schema to not throw within `subscribe`? Is this a possibility? If not, what meaning does the error in `subscribe` have?

**Reply by elderapo** [reply]
Fixing the schema to not throw within a `subscribe` seems like a good idea.

**Reply by RMHonor** [reply]
A warning to anyone using our solution above (https://github.com/enisdenjo/graphql-ws/discussions/333#discussioncomment-2332628), it seems there's a memory leak where the request context persists. Just investigating now to see what's holding onto the memory.

**Reply by RMHonor** [reply]
Found the issue. When an error is thrown, the following line is not reached: https://github.com/enisdenjo/graphql-ws/blob/45d1649474a7e95b0b1f6530e2f3d3cb3d734db8/src/server.ts#L822
This means that the `ctx.subscriptions` map continues to grow over time.

A workaround we've got for now is send a `MessageType.Complete` to the handler, which cleans up the subscription.

Perhaps the above line could be wrapped up in a `try/finally`, so the subscription is deleted?

**Reply by enisdenjo** [reply]
Great catch @RMHonor, thanks for pointing it out! I'll take care of it as soon as possible.

**Reply by enisdenjo** [reply]
Hey @RMHonor, fix for https://github.com/enisdenjo/graphql-ws/discussions/333#discussioncomment-2738402 is now live with [v5.8.2](https://github.com/enisdenjo/graphql-ws/releases/tag/v5.8.2), try it out! 😄

**Reply by RMHonor** [reply]
Wow, quick work @enisdenjo, thanks a lot!

**Reply by aseemk** [reply]
For anyone else interested, I opened this issue with `graphql-js` to hopefully change this behavior:

https://github.com/graphql/graphql-js/issues/4001

+1 reactions and/or any additional perspectives would be welcome! Thanks.

**Comment by enisdenjo** [comment, marked as answer]
Please refer to the recipe showcasing how someone may [implement a custom subscribe for graceful handling thrown errors](https://the-guild.dev/graphql/ws/recipes#ws-server-usage-with-custom-subscribe-method-that-gracefully-handles-thrown-errors) with just 15 lines of code.

**Reply by ralphstodomingo** [reply]
Just wanted to chime in - I followed that recipe, but it infinitely calls result.next, until the client disconnects, where I get a `write EPIPE` error.

In my case, trying to put `done: true` in the object returned within catch to stop the infinite looping, but TS expects me to return just a value field like the recipe. Also, the error doesn't balloon back to the client that way.

Another approach I tried was to return `result.throw` to let the async iterator that threw the error to clean up, but `graphql-redis-subscriptions` throws an error anyway in their async-iterator throw fn, so that still outputs `Internal error occurred during message handling. Please check your implementation.` for me, but at least it still lets the client to connect to the server that way.

**Reply by enisdenjo** [reply]
Hey! So, the thing is, `.next` can throw and have the iterator continue. The iterator stopping when `.next` throws is an assumption and hence an implementation detail. You're expected to implement a mechanism that breaks the iterator when `.next` throws yourself.

**Reply by ralphstodomingo** [reply]
I see! I'm not very familiar with the use of async iterators here, but what I did, where I had to invoke .throw manually in the recipe's catch is a valid way to have the iterator stop there as needed?
